contrib/redis/go-redis.v9/redis.go

```diff
@@ -0,0 +1,185 @@
+package redis
+
+import (
+	"bytes"
+	"context"
+	"math"
+	"net"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/redis/go-redis/v9"
+)
+
+type datadogHook struct {
+	*params
+}
+
+// params holds the tracer and a set of parameters which are recorded with every trace.
+type params struct {
+	config         *clientConfig
+	additionalTags []ddtrace.StartSpanOption
+}
+
+// NewClient returns a new Client that is traced with the default tracer under
+// the service name "redis".
+func NewClient(opt *redis.Options, opts ...ClientOption) redis.UniversalClient {
+	client := redis.NewClient(opt)
+	WrapClient(client, opts...)
+	return client
+}
+
+// WrapClient adds a hook to the given client that traces with the default tracer under
+// the service name "redis".
+func WrapClient(client redis.UniversalClient, opts ...ClientOption) {
+	cfg := new(clientConfig)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+
+	hookParams := &params{
+		additionalTags: additionalTagOptions(client),
+		config:         cfg,
+	}
+
+	client.AddHook(&datadogHook{params: hookParams})
+}
+
+type clientOptions interface {
+	Options() *redis.Options
+}
+
+type clusterOptions interface {
+	Options() *redis.ClusterOptions
+}
+
+func additionalTagOptions(client redis.UniversalClient) []ddtrace.StartSpanOption {
+	additionalTags := []ddtrace.StartSpanOption{}
+	if clientOptions, ok := client.(clientOptions); ok {
+		opt := clientOptions.Options()
+		if opt.Addr == "FailoverClient" {
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		} else {
+			host, port, err := net.SplitHostPort(opt.Addr)
+			if err != nil {
+				host = opt.Addr
+				port = "6379"
+			}
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag(ext.TargetHost, host),
+				tracer.Tag(ext.TargetPort, port),
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		}
+	} else if clientOptions, ok := client.(clusterOptions); ok {
+		addrs := []string{}
+		for _, addr := range clientOptions.Options().Addrs {
+			addrs = append(addrs, addr)
+		}
+		additionalTags = []ddtrace.StartSpanOption{
+			tracer.Tag("addrs", strings.Join(addrs, ", ")),
+		}
+	}
+	return additionalTags
+}
+
+func (ddh *datadogHook) DialHook(hook redis.DialHook) redis.DialHook {
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		conn, err := hook(ctx, network, addr)
+		return conn, err
+	}
+}
+
+func (ddh *datadogHook) ProcessHook(hook redis.ProcessHook) redis.ProcessHook {
+	return func(ctx context.Context, cmd redis.Cmder) error {
+		raw := cmd.String()
+		length := strings.Count(raw, " ")
+		p := ddh.params
+		startOpts := make([]ddtrace.StartSpanOption, 0, 4+1+len(ddh.additionalTags)+1) // 4 options below + redis.raw_command + ddh.additionalTags + analyticsRate
```

Comment:
Unless I am misunderstanding, I count seven options below, not four? So we should use 7 instead of 4 here.

Reply:
You're right, it should be 7 instead of 4. 
I'll update the code. Thank you!

This part of the code is directly taken from v8 code, so it seems the capacity update was missed when the options increased previously.

Reply:
fix in [5d4e5c6](https://github.com/DataDog/dd-trace-go/pull/1730/commits/5d4e5c6d218e5f39fa67b320219fb30f27601cc1)